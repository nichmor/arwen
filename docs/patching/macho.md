## Understanding Mach-O Files and Patching

Mach-O (Mach Object) is the native binary file format used by Apple operating systems, including macOS, iOS, iPadOS, tvOS, and watchOS. It defines the structure for executables, dynamic libraries (`.dylib`), frameworks, bundles, and object files (`.o`). Understanding its structure is key to manipulating dependencies and runtime behavior, often necessary for application packaging and relocation.

### Mach-O File Structure

A Mach-O file is organized into three main regions:

- **Mach Header:** Located at the very beginning (offset 0). It identifies the file as Mach-O, specifies the target architecture (e.g., x86_64, arm64), the file type (executable, dylib, etc.), and most importantly, the number and total size of the Load Commands that follow.
- **Load Commands:** A list of variable-length commands immediately following the header. These commands act as instructions for the operating system's dynamic linker (`dyld`) and the kernel, dictating how to map the file into memory, what libraries are needed, where the main execution thread starts, symbol table locations, code signature details, and more.
- **Data:** The bulk of the file, containing the actual code and data, organized into segments and sections as specified by the Load Commands.

**Visual Layout**

```plaintext
+---------------------+
| Mach Header         | (File Offset 0)
| (mach_header_64)    |
+---------------------+
| Load Command 1      | (Immediately follows Header)
+---------------------+
| Load Command 2      |
+---------------------+
| ...                 |
+---------------------+
| Load Command N      | (Total size = sizeofcmds from Header)
+---------------------+
|                     |
| Data Region 1       | (e.g., __TEXT Segment: code sections)
| (Segments/Sections) |
+---------------------+
|                     |
| Data Region 2       | (e.g., __DATA Segment: data sections)
|                     |
+---------------------+
| ...                 | (e.g., __LINKEDIT: symbol/string tables)
+---------------------+
```

### Load Commands: The Core Instructions

The Load Commands region is central to Mach-O's functionality. `dyld` parses this list to understand how to prepare the binary for execution. Each command has a type (`cmd`) and size (`cmdsize`). Key types include:

- **`LC_SEGMENT_64`** / **`LC_SEGMENT`**: Defines a segment (e.g., `__TEXT`, `__DATA`, `__LINKEDIT`) and its properties: file offset/size, virtual memory address/size, and permissions (read/write/execute). It also contains descriptions of the sections (like `__TEXT.__text`, `__DATA.__data`) within that segment.
- **`LC_ID_DYLIB`**: Specifies the "install name" for a dynamic library. This is the canonical path identifying the library, used by other binaries when linking against it.
- **`LC_LOAD_DYLIB`**: Defines a dependency on an external dynamic library, specifying the library's install name (the path to find it).
- **`LC_LOAD_WEAK_DYLIB`**: Defines an optional library dependency.
- **`LC_REEXPORT_DYLIB`**: Links against another library and re-exports its symbols.
- **`LC_RPATH`**: Adds a path to the runtime search path list, used for resolving `@rpath` dependencies.
- **`LC_MAIN`**: Specifies the entry point (start address) for executable files.
- **`LC_CODE_SIGNATURE`**: Points to the code signature data.
- **`LC_SYMTAB`**: Points to the symbol table and string table (used by linker/debugger).
- **`LC_DYSYMTAB`**: Points to dynamic linking symbol information.
- **`LC_DYLD_INFO_ONLY`**: Points to optimized dynamic linking info used by `dyld` (rebasing, binding).

### Path Commands and Resolution

How `dyld` finds dependent libraries (`LC_LOAD_DYLIB`) is crucial and often involves special path prefixes:

- **`@executable_path`**: Resolves to the absolute path of the directory containing the main executable of the running process. Useful for finding libraries bundled *relative* to the main application binary.
- **`@loader_path`**: Resolves to the absolute path of the directory containing the specific Mach-O file (executable *or* library) that contains the `LC_LOAD_DYLIB` command currently being processed. Useful for libraries finding other libraries located relative to themselves.
- **`@rpath`**: A placeholder indicating that `dyld` should search for the library using a list of runtime search paths. This search list is constructed in order:
    -  Paths specified by `LC_RPATH` load commands within the Mach-O file containing the `@rpath` dependency itself.
    -  Paths specified by `LC_RPATH` load commands within the main executable (if the dependency is not in the main executable).
    -  Paths specified by `LC_RPATH` load commands within the main executable (if the dependency is not in the main executable).
    -  Paths specified in the `DYLD_LIBRARY_PATH` environment variable (though its use is often restricted for security reasons, especially with System Integrity Protection).
    -  Paths specified in the `DYLD_FALLBACK_LIBRARY_PATH` environment variable (if `DYLD_LIBRARY_PATH` is not set or doesn't find the library).
    -  Standard system fallback locations (e.g., `/usr/local/lib`, `/usr/lib`).

The `LC_RPATH` load command simply contains a path string to be added to this search list.

### How Patching Works with `arwen`

Patching Mach-O files with a tool like `arwen` typically involves modifying the Load Commands or the data they reference (often strings within the commands themselves or in the `__LINKEDIT` segment).

Some common patching operations include:

- **Modifying runtime dependencies**.
This involves adding `LC_RPATH` or removing them. For example, adding `LC_RPATH` command with the path `@loader_path/../Frameworks` to make a binary look inside a sibling `Frameworks` directory for its `@rpath` dependencies.

- **Changing Dependencies:** To make a binary look for a library in a different location, you modify the path string stored within an `LC_LOAD_DYLIB` or `LC_LOAD_WEAK_DYLIB` command. For example, changing `/usr/local/lib/libfoo.dylib` to `@rpath/libfoo.dylib` often requires ensuring an appropriate `LC_RPATH` exists.

- **Changing a Library's Install Name:** To change the canonical path by which other binaries refer to a library (essential when relocating or bundling libraries/frameworks), you modify the path string stored within the library's own `LC_ID_DYLIB` command. For instance, changing `/Users/dev/project/build/lib/libbar.dylib` to `@rpath/libbar.dylib`.

- **Adding or Modifying Runtime Search Paths (RPATH):** To tell `dyld` where to look when resolving `@rpath` dependencies, you add a new `LC_RPATH` command or modify the path string within an existing one. You might add an `LC_RPATH` command with the path `@loader_path/../Frameworks` to make a binary look inside a sibling `Frameworks` directory for its `@rpath` dependencies.

**Challenges and Considerations:**

* **Space Constraints:** The `mach_header` specifies the total size (`sizeofcmds`) allocated for all load commands. If you need to add a new command or make a path string significantly longer, there might not be enough space. Simple tools might fail. More sophisticated tools like `arwen` might attempt to use existing padding or might need to rewrite parts of the file, which is complex. Changing a path to another path of the *same or shorter length* is generally safest and easiest.
* **Code Signing:** Modifying *any* part of a signed Mach-O binary (executable or library) **invalidates its code signature**. On modern macOS and iOS, unsigned or improperly signed code may fail to run due to security policies (Gatekeeper, System Integrity Protection). After patching a signed binary, you **must re-sign it** using the `codesign` command-line tool with an appropriate certificate for it to be runnable in many contexts.
