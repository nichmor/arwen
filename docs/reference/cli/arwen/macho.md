<!--- This file is autogenerated. Do not edit manually! -->
# <code>[arwen](../arwen.md) macho</code>

## About
Mach-O commands

--8<-- "docs/reference/cli/arwen/macho_extender.md:description"

## Usage
```
arwen macho <COMMAND>
```

## Subcommands
| Command | Description |
|---------|-------------|
| [`delete-rpath`](macho/delete-rpath.md) | Remove a run path |
| [`change-rpath`](macho/change-rpath.md) | Change already existing run path |
| [`add-rpath`](macho/add-rpath.md) | Add a run path |
| [`change-install-name`](macho/change-install-name.md) | Change existing dylib load name |
| [`change-install-id`](macho/change-install-id.md) | Change dylib id. Works only if your object is a shared library |


--8<-- "docs/reference/cli/arwen/macho_extender.md:example"
