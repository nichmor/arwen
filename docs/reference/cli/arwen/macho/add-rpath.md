<!--- This file is autogenerated. Do not edit manually! -->
# <code>[arwen](../../arwen.md) [macho](../macho.md) add-rpath</code>

## About
Add a run path

--8<-- "docs/reference/cli/arwen/macho/add-rpath_extender.md:description"

## Usage
```
arwen macho add-rpath <NEW_RPATH> <PATH_TO_BINARY>
```

## Arguments
- <a id="arg-<NEW_RPATH>" href="#arg-<NEW_RPATH>">`<NEW_RPATH>`</a>
:  New rpath to add
<br>**required**: `true`
- <a id="arg-<PATH_TO_BINARY>" href="#arg-<PATH_TO_BINARY>">`<PATH_TO_BINARY>`</a>
:  Path to the file to change
<br>**required**: `true`

--8<-- "docs/reference/cli/arwen/macho/add-rpath_extender.md:example"
