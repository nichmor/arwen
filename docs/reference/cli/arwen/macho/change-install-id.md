<!--- This file is autogenerated. Do not edit manually! -->
# <code>[arwen](../../arwen.md) [macho](../macho.md) change-install-id</code>

## About
Change dylib id. Works only if your object is a shared library

--8<-- "docs/reference/cli/arwen/macho/change-install-id_extender.md:description"

## Usage
```
arwen macho change-install-id <NEW_INSTALL_ID> <PATH_TO_BINARY>
```

## Arguments
- <a id="arg-<NEW_INSTALL_ID>" href="#arg-<NEW_INSTALL_ID>">`<NEW_INSTALL_ID>`</a>
:  New rpath to add
<br>**required**: `true`
- <a id="arg-<PATH_TO_BINARY>" href="#arg-<PATH_TO_BINARY>">`<PATH_TO_BINARY>`</a>
:  Path to the file to change
<br>**required**: `true`

--8<-- "docs/reference/cli/arwen/macho/change-install-id_extender.md:example"
