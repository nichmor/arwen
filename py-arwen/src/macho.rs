use arwen::macho::MachoContainer;
use pyo3::exceptions::PyException;
use pyo3::prelude::*;
use std::fs;
use std::path::Path;

use crate::error::PyMachoError;

/// Python wrapper for Rust's MachoContainer
#[pyclass]
#[repr(transparent)]
pub struct PyMachoContainer {
    inner: MachoContainer<'static>,
}

#[pymethods]
impl PyMachoContainer {
    /// Create a new MachoContainer by reading a file from a path
    #[staticmethod]
    #[pyo3(text_signature = "(path)")]
    fn from_path(path: &str) -> PyResult<Self> {
        let file_path = Path::new(path);
        let data = fs::read(file_path).map_err(|e| {
            PyErr::new::<PyException, _>(format!("Failed to read file {}: {}", path, e))
        })?;

        // Note: This is safe because we're storing the data in the struct
        let static_data = Box::leak(data.into_boxed_slice());

        let container = MachoContainer::parse(static_data).map_err(PyMachoError::from)?;

        Ok(PyMachoContainer { inner: container })
    }

    /// Add a new rpath to the Mach-O file
    #[pyo3(text_signature = "($self, new_rpath)")]
    fn add_rpath(&mut self, new_rpath: &str) -> PyResult<()> {
        Ok(self
            .inner
            .add_rpath(new_rpath)
            .map_err(PyMachoError::from)?)
    }

    /// Change an existing rpath in the Mach-O file
    #[pyo3(text_signature = "($self, old_rpath, new_rpath)")]
    fn change_rpath(&mut self, old_rpath: &str, new_rpath: &str) -> PyResult<()> {
        Ok(self
            .inner
            .change_rpath(old_rpath, new_rpath)
            .map_err(PyMachoError::from)?)
    }

    /// Remove an existing rpath from the Mach-O file
    #[pyo3(text_signature = "($self, old_rpath)")]
    fn remove_rpath(&mut self, old_rpath: &str) -> PyResult<()> {
        Ok(self
            .inner
            .remove_rpath(old_rpath)
            .map_err(PyMachoError::from)?)
    }

    /// Change the install ID of the Mach-O file (for shared libraries)
    #[pyo3(text_signature = "($self, new_id)")]
    fn change_install_id(&mut self, new_id: &str) -> PyResult<()> {
        Ok(self
            .inner
            .change_install_id(new_id)
            .map_err(PyMachoError::from)?)
    }

    /// Change the install name of a dependency in the Mach-O file
    #[pyo3(text_signature = "($self, old_name, new_name)")]
    fn change_install_name(&mut self, old_name: &str, new_name: &str) -> PyResult<()> {
        Ok(self
            .inner
            .change_install_name(old_name, new_name)
            .map_err(PyMachoError::from)?)
    }

    /// Save the modified Mach-O file to a path
    #[pyo3(text_signature = "($self, path)")]
    fn save(&self, path: &str) -> PyResult<()> {
        fs::write(path, &self.inner.data).map_err(|e| {
            PyErr::new::<PyException, _>(format!("Failed to write file {}: {}", path, e))
        })
    }
}
