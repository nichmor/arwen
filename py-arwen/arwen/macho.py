"""
Mach-O binary manipulation module for macOS.

This module provides functionality for working with Mach-O binary files,
allowing operations like modifying rpaths, changing install names, and more.
"""

from arwen.arwen import PyMachoContainer


class MachoContainer:
    """
    Python wrapper for manipulating Mach-O binary files.

    This class provides methods to modify various aspects of Mach-O binaries,
    such as rpaths, install names, and install IDs.
    """

    _inner: PyMachoContainer

    @staticmethod
    def from_path(path: str) -> "MachoContainer":
        """
        Create a new MachoContainer by parsing a Mach-O binary file.

        Args:
            path: Path to the Mach-O binary file

        Returns:
            A new MachoContainer instance

        Raises:
            MachoError: If the file cannot be parsed as a Mach-O binary
            Exception: If the file cannot be read
        """
        return MachoContainer._from_py(PyMachoContainer.from_path(path))

    def add_rpath(self, new_rpath: str) -> None:
        """
        Add a new rpath to the Mach-O file.

        Args:
            new_rpath: The rpath to add

        Raises:
            MachoError: If the rpath cannot be added
        """
        self._inner.add_rpath(new_rpath)

    def change_rpath(self, old_rpath: str, new_rpath: str) -> None:
        """
        Change an existing rpath in the Mach-O file.

        Args:
            old_rpath: The existing rpath to change
            new_rpath: The new rpath value

        Raises:
            RpathMissingError: If the old_rpath doesn't exist
            MachoError: If the rpath cannot be changed
        """
        self._inner.change_rpath(old_rpath, new_rpath)

    def remove_rpath(self, old_rpath: str) -> None:
        """
        Remove an existing rpath from the Mach-O file.

        Args:
            old_rpath: The rpath to remove

        Raises:
            RpathMissingError: If the specified rpath doesn't exist
            MachoError: If the rpath cannot be removed
        """
        self._inner.remove_rpath(old_rpath)

    def change_install_id(self, new_id: str) -> None:
        """
        Change the install ID of the Mach-O file (for shared libraries).

        Args:
            new_id: The new install ID

        Raises:
            DylibIdMissingError: If the file is not a shared library
            MachoError: If the install ID cannot be changed
        """
        self._inner.change_install_id(new_id)

    def change_install_name(self, old_name: str, new_name: str) -> None:
        """
        Change the install name of a dependency in the Mach-O file.

        Args:
            old_name: The existing install name to change
            new_name: The new install name

        Raises:
            DylibNameMissingError: If the old_name doesn't exist
            MachoError: If the install name cannot be changed
        """
        self._inner.change_install_name(old_name, new_name)

    def save(self, path: str) -> None:
        """
        Save the modified Mach-O file to a path.

        Args:
            path: Path where the modified file will be saved

        Raises:
            Exception: If the file cannot be written
        """
        self._inner.save(path)

    @classmethod
    def _from_py(cls, py_macho: PyMachoContainer) -> "MachoContainer":
        macho_container = cls.__new__(cls)
        macho_container._inner = py_macho

        return macho_container
