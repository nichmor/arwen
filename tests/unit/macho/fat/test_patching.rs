/// This test checks if the rpath of a Mach-O file can be changed.
use arwen::macho::MachoContainer;
use goblin::mach::{MultiArch, SingleArch};
use rstest::rstest;
use std::path::PathBuf;

// const FAT_BINARY_PATH: &str = "tests/data/macho/fat/i386_x86_64/hello_with_some_rpath.bin";
// const FAT_LIB_PATH: &str = "tests/data/macho/fat/x64_x86_64/fatlib";

#[rstest]
fn test_rpath_change(#[files("tests/data/macho/fat/exec/*")] bin_path: PathBuf) {
    // load hello_with_rpath from data

    let data_bytes = std::fs::read(&bin_path).unwrap();

    let mut macho_container = MachoContainer::parse(&data_bytes).unwrap();

    macho_container
        .change_rpath("path_graf", "path_graf_path_graf_path_graf_path_graf")
        .unwrap();

    let fat_macho = MultiArch::new(&macho_container.data).unwrap();

    for arch in fat_macho.into_iter() {
        let single_arch = arch.unwrap();

        let SingleArch::MachO(changed_macho) = single_arch else {
            panic!("Expected MachO, got {:?}", single_arch);
        };

        // let changed_macho = MachO::parse(&arch.data, 0).unwrap();
        insta::assert_debug_snapshot!(changed_macho);
        insta::assert_debug_snapshot!(changed_macho.rpaths);
    }

    insta::assert_snapshot!(macho_container.data.len(), @"76224");
}

/// This test checks if the rpath of a Mach-O file can be removed.
#[rstest]
fn test_rpath_remove(#[files("tests/data/macho/fat/exec/*")] bin_path: PathBuf) {
    let data_bytes = std::fs::read(&bin_path).unwrap();

    let mut macho_container = MachoContainer::parse(&data_bytes).unwrap();

    // we try to change the rpath of the file to something longer
    macho_container.remove_rpath("path_graf").unwrap();

    let fat_macho = MultiArch::new(&macho_container.data).unwrap();

    for arch in fat_macho.into_iter() {
        let single_arch = arch.unwrap();

        let SingleArch::MachO(changed_macho) = single_arch else {
            panic!("Expected MachO, got {:?}", single_arch);
        };

        // let changed_macho = MachO::parse(&arch.data, 0).unwrap();
        insta::assert_debug_snapshot!(changed_macho);
        insta::assert_debug_snapshot!(changed_macho.rpaths);
    }

    insta::assert_snapshot!(macho_container.data.len());
}

/// This test verify that a rpath can be added to a Mach-O file.
#[rstest]
fn test_rpath_add(#[files("tests/data/macho/fat/exec/*")] bin_path: PathBuf) {
    // load hello_with_rpath from data
    let data_bytes = std::fs::read(&bin_path).unwrap();

    let mut macho_container = MachoContainer::parse(&data_bytes).unwrap();

    // we try to change the rpath of the file to something longer
    macho_container
        .add_rpath("abababababababababababaabbababababababababab")
        .unwrap();

    let fat_macho = MultiArch::new(&macho_container.data).unwrap();

    for arch in fat_macho.into_iter() {
        let single_arch = arch.unwrap();

        let SingleArch::MachO(changed_macho) = single_arch else {
            panic!("Expected MachO, got {:?}", single_arch);
        };

        // let changed_macho = MachO::parse(&arch.data, 0).unwrap();
        insta::assert_debug_snapshot!(changed_macho);
        insta::assert_debug_snapshot!(changed_macho.rpaths);
    }

    insta::assert_snapshot!(macho_container.data.len());
}

/// This test verify that an dylib id can be changed in a Mach-O file.
#[rstest]
fn test_change_dylib_id(#[files("tests/data/macho/fat/libs/*")] lib_path: PathBuf) {
    let data_bytes = std::fs::read(&lib_path).unwrap();

    let mut macho_container = MachoContainer::parse(&data_bytes).unwrap();

    // we try to change the rpath of the file to something longer
    macho_container
        .change_install_id("very_very_very_very_very_very_very_longid.dylib")
        .unwrap();

    let fat_macho = MultiArch::new(&macho_container.data).unwrap();

    for arch in fat_macho.into_iter() {
        let single_arch = arch.unwrap();

        let SingleArch::MachO(changed_macho) = single_arch else {
            panic!("Expected MachO, got {:?}", single_arch);
        };

        // let changed_macho = MachO::parse(&arch.data, 0).unwrap();
        insta::assert_debug_snapshot!(changed_macho);
        insta::assert_debug_snapshot!(changed_macho.name);
    }

    insta::assert_snapshot!(macho_container.data.len());
}

/// This test verify that a name of dynamic library can be changed
#[rstest]
fn test_change_dylib_name(#[files("tests/data/macho/fat/exec/*")] bin_path: PathBuf) {
    let data_bytes = std::fs::read(&bin_path).unwrap();

    let mut macho_container = MachoContainer::parse(&data_bytes).unwrap();

    // we try to change the rpath of the file to something longer
    macho_container
        .change_install_name(
            "/usr/lib/libSystem.B.dylib",
            "very_very_very_very_very_very_very_longid",
        )
        .unwrap();

    let fat_macho = MultiArch::new(&macho_container.data).unwrap();

    for arch in fat_macho.into_iter() {
        let single_arch = arch.unwrap();

        let SingleArch::MachO(changed_macho) = single_arch else {
            panic!("Expected MachO, got {:?}", single_arch);
        };

        // let changed_macho = MachO::parse(&arch.data, 0).unwrap();
        insta::assert_debug_snapshot!(changed_macho);
        insta::assert_debug_snapshot!(changed_macho.libs);
    }

    insta::assert_snapshot!(macho_container.data.len());
}
