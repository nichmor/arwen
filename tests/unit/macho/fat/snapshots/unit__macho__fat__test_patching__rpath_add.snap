---
source: tests/unit/macho/fat/test_patching.rs
expression: changed_macho
---
MachO {
    header: Header {
        magic: 0xfeedface,
        cputype: 7,
        cpusubtype: 0x3,
        filetype: "EXECUTE",
        ncmds: 19,
        sizeofcmds: 1172,
        flags: 0x1200085,
        reserved: 0x0,
    },
    load_commands: [
        LoadCommand {
            offset: 28,
            command: Segment32(
                SegmentCommand32 {
                    cmd: 1,
                    cmdsize: 56,
                    segname: [
                        95,
                        95,
                        80,
                        65,
                        71,
                        69,
                        90,
                        69,
                        82,
                        79,
                        0,
                        0,
                        0,
                        0,
                        0,
                        0,
                    ],
                    vmaddr: 0,
                    vmsize: 4096,
                    fileoff: 0,
                    filesize: 0,
                    maxprot: 0,
                    initprot: 0,
                    nsects: 0,
                    flags: 0,
                },
            ),
        },
        LoadCommand {
            offset: 84,
            command: Segment32(
                SegmentCommand32 {
                    cmd: 1,
                    cmdsize: 396,
                    segname: [
                        95,
                        95,
                        84,
                        69,
                        88,
                        84,
                        0,
                        0,
                        0,
                        0,
                        0,
                        0,
                        0,
                        0,
                        0,
                        0,
                    ],
                    vmaddr: 4096,
                    vmsize: 4096,
                    fileoff: 0,
                    filesize: 4096,
                    maxprot: 7,
                    initprot: 5,
                    nsects: 5,
                    flags: 0,
                },
            ),
        },
        LoadCommand {
            offset: 480,
            command: Segment32(
                SegmentCommand32 {
                    cmd: 1,
                    cmdsize: 192,
                    segname: [
                        95,
                        95,
                        68,
                        65,
                        84,
                        65,
                        0,
                        0,
                        0,
                        0,
                        0,
                        0,
                        0,
                        0,
                        0,
                        0,
                    ],
                    vmaddr: 8192,
                    vmsize: 4096,
                    fileoff: 4096,
                    filesize: 4096,
                    maxprot: 7,
                    initprot: 3,
                    nsects: 2,
                    flags: 0,
                },
            ),
        },
        LoadCommand {
            offset: 672,
            command: Segment32(
                SegmentCommand32 {
                    cmd: 1,
                    cmdsize: 56,
                    segname: [
                        95,
                        95,
                        76,
                        73,
                        78,
                        75,
                        69,
                        68,
                        73,
                        84,
                        0,
                        0,
                        0,
                        0,
                        0,
                        0,
                    ],
                    vmaddr: 12288,
                    vmsize: 32768,
                    fileoff: 8192,
                    filesize: 18864,
                    maxprot: 7,
                    initprot: 1,
                    nsects: 0,
                    flags: 0,
                },
            ),
        },
        LoadCommand {
            offset: 728,
            command: DyldInfoOnly(
                DyldInfoCommand {
                    cmd: 2147483682,
                    cmdsize: 48,
                    rebase_off: 8192,
                    rebase_size: 16,
                    bind_off: 8208,
                    bind_size: 24,
                    weak_bind_off: 0,
                    weak_bind_size: 0,
                    lazy_bind_off: 8232,
                    lazy_bind_size: 16,
                    export_off: 8248,
                    export_size: 44,
                },
            ),
        },
        LoadCommand {
            offset: 776,
            command: Symtab(
                SymtabCommand {
                    cmd: 2,
                    cmdsize: 24,
                    symoff: 8296,
                    nsyms: 4,
                    stroff: 8360,
                    strsize: 56,
                },
            ),
        },
        LoadCommand {
            offset: 800,
            command: Dysymtab(
                DysymtabCommand {
                    cmd: 11,
                    cmdsize: 80,
                    ilocalsym: 0,
                    nlocalsym: 0,
                    iextdefsym: 0,
                    nextdefsym: 2,
                    iundefsym: 2,
                    nundefsym: 2,
                    tocoff: 0,
                    ntoc: 0,
                    modtaboff: 0,
                    nmodtab: 0,
                    extrefsymoff: 0,
                    nextrefsyms: 0,
                    indirectsymoff: 8344,
                    nindirectsyms: 4,
                    extreloff: 0,
                    nextrel: 0,
                    locreloff: 0,
                    nlocrel: 0,
                },
            ),
        },
        LoadCommand {
            offset: 880,
            command: LoadDylinker(
                DylinkerCommand {
                    cmd: 14,
                    cmdsize: 28,
                    name: 12,
                },
            ),
        },
        LoadCommand {
            offset: 908,
            command: Uuid(
                UuidCommand {
                    cmd: 27,
                    cmdsize: 24,
                    uuid: [
                        201,
                        149,
                        203,
                        167,
                        55,
                        135,
                        53,
                        19,
                        187,
                        35,
                        96,
                        89,
                        116,
                        221,
                        54,
                        9,
                    ],
                },
            ),
        },
        LoadCommand {
            offset: 932,
            command: VersionMinMacosx(
                VersionMinCommand {
                    cmd: 36,
                    cmdsize: 16,
                    version: 658176,
                    sdk: 658176,
                },
            ),
        },
        LoadCommand {
            offset: 948,
            command: SourceVersion(
                SourceVersionCommand {
                    cmd: 42,
                    cmdsize: 16,
                    version: 0,
                },
            ),
        },
        LoadCommand {
            offset: 964,
            command: Main(
                EntryPointCommand {
                    cmd: 2147483688,
                    cmdsize: 24,
                    entryoff: 3920,
                    stacksize: 0,
                },
            ),
        },
        LoadCommand {
            offset: 988,
            command: LoadDylib(
                DylibCommand {
                    cmd: 12,
                    cmdsize: 52,
                    dylib: Dylib {
                        name: 24,
                        timestamp: 2,
                        current_version: 80281857,
                        compatibility_version: 65536,
                    },
                },
            ),
        },
        LoadCommand {
            offset: 1040,
            command: Rpath(
                RpathCommand {
                    cmd: 2147483676,
                    cmdsize: 28,
                    path: 12,
                },
            ),
        },
        LoadCommand {
            offset: 1068,
            command: FunctionStarts(
                LinkeditDataCommand {
                    cmd: 38,
                    cmdsize: 16,
                    dataoff: 8292,
                    datasize: 4,
                },
            ),
        },
        LoadCommand {
            offset: 1084,
            command: DataInCode(
                LinkeditDataCommand {
                    cmd: 41,
                    cmdsize: 16,
                    dataoff: 8296,
                    datasize: 0,
                },
            ),
        },
        LoadCommand {
            offset: 1100,
            command: CodeSignature(
                LinkeditDataCommand {
                    cmd: 29,
                    cmdsize: 16,
                    dataoff: 8416,
                    datasize: 18640,
                },
            ),
        },
        LoadCommand {
            offset: 1116,
            command: Rpath(
                RpathCommand {
                    cmd: 2147483676,
                    cmdsize: 24,
                    path: 12,
                },
            ),
        },
        LoadCommand {
            offset: 1140,
            command: Rpath(
                RpathCommand {
                    cmd: 2147483676,
                    cmdsize: 60,
                    path: 12,
                },
            ),
        },
    ],
    segments: Segments {
        segments: [
            Segment {
                cmd: 1,
                cmdsize: 56,
                segname: "__PAGEZERO",
                vmaddr: 0,
                vmsize: 4096,
                fileoff: 0,
                filesize: 0,
                maxprot: 0,
                initprot: 0,
                nsects: 0,
                flags: 0,
                data: 0,
                sections(): Ok(
                    [],
                ),
            },
            Segment {
                cmd: 1,
                cmdsize: 396,
                segname: "__TEXT",
                vmaddr: 4096,
                vmsize: 4096,
                fileoff: 0,
                filesize: 4096,
                maxprot: 7,
                initprot: 5,
                nsects: 5,
                flags: 0,
                data: 4096,
                sections(): Ok(
                    [
                        Section {
                            sectname: "__text",
                            segname: "__TEXT",
                            addr: 8016,
                            size: 50,
                            offset: 3920,
                            align: 4,
                            reloff: 0,
                            nreloc: 0,
                            flags: 2147484672,
                        },
                        Section {
                            sectname: "__symbol_stub",
                            segname: "__TEXT",
                            addr: 8066,
                            size: 6,
                            offset: 3970,
                            align: 1,
                            reloff: 0,
                            nreloc: 0,
                            flags: 2147484936,
                        },
                        Section {
                            sectname: "__stub_helper",
                            segname: "__TEXT",
                            addr: 8072,
                            size: 22,
                            offset: 3976,
                            align: 2,
                            reloff: 0,
                            nreloc: 0,
                            flags: 2147484928,
                        },
                        Section {
                            sectname: "__cstring",
                            segname: "__TEXT",
                            addr: 8094,
                            size: 15,
                            offset: 3998,
                            align: 0,
                            reloff: 0,
                            nreloc: 0,
                            flags: 2,
                        },
                        Section {
                            sectname: "__unwind_info",
                            segname: "__TEXT",
                            addr: 8112,
                            size: 72,
                            offset: 4016,
                            align: 2,
                            reloff: 0,
                            nreloc: 0,
                            flags: 0,
                        },
                    ],
                ),
            },
            Segment {
                cmd: 1,
                cmdsize: 192,
                segname: "__DATA",
                vmaddr: 8192,
                vmsize: 4096,
                fileoff: 4096,
                filesize: 4096,
                maxprot: 7,
                initprot: 3,
                nsects: 2,
                flags: 0,
                data: 4096,
                sections(): Ok(
                    [
                        Section {
                            sectname: "__nl_symbol_ptr",
                            segname: "__DATA",
                            addr: 8192,
                            size: 8,
                            offset: 4096,
                            align: 2,
                            reloff: 0,
                            nreloc: 0,
                            flags: 6,
                        },
                        Section {
                            sectname: "__la_symbol_ptr",
                            segname: "__DATA",
                            addr: 8200,
                            size: 4,
                            offset: 4104,
                            align: 2,
                            reloff: 0,
                            nreloc: 0,
                            flags: 7,
                        },
                    ],
                ),
            },
            Segment {
                cmd: 1,
                cmdsize: 56,
                segname: "__LINKEDIT",
                vmaddr: 12288,
                vmsize: 32768,
                fileoff: 8192,
                filesize: 18864,
                maxprot: 7,
                initprot: 1,
                nsects: 0,
                flags: 0,
                data: 18864,
                sections(): Ok(
                    [],
                ),
            },
        ],
    },
    entry: 8016,
    old_style_entry: false,
    libs: [
        "self",
        "/usr/lib/libSystem.B.dylib",
    ],
    name: None,
    little_endian: true,
    is_64: false,
    symbols(): [
        Ok(
            (
                "__mh_execute_header",
                Nlist {
                    n_strx: 2,
                    n_type: 15,
                    n_sect: 1,
                    n_desc: 16,
                    n_value: 4096,
                },
            ),
        ),
        Ok(
            (
                "_main",
                Nlist {
                    n_strx: 22,
                    n_type: 15,
                    n_sect: 1,
                    n_desc: 0,
                    n_value: 8016,
                },
            ),
        ),
        Ok(
            (
                "_printf",
                Nlist {
                    n_strx: 28,
                    n_type: 1,
                    n_sect: 0,
                    n_desc: 256,
                    n_value: 0,
                },
            ),
        ),
        Ok(
            (
                "dyld_stub_binder",
                Nlist {
                    n_strx: 36,
                    n_type: 1,
                    n_sect: 0,
                    n_desc: 256,
                    n_value: 0,
                },
            ),
        ),
    ],
    exports(): Ok(
        [
            Export {
                name: "__mh_execute_header",
                info: Regular {
                    address: 0,
                    flags: 0,
                },
                size: 0,
                offset: 0,
            },
            Export {
                name: "_main",
                info: Regular {
                    address: 3920,
                    flags: 0,
                },
                size: 0,
                offset: 3920,
            },
        ],
    ),
    imports(): Ok(
        [
            Import {
                name: "dyld_stub_binder",
                dylib: "/usr/lib/libSystem.B.dylib",
                is_lazy: false,
                offset: 4096,
                size: 0,
                address: 8192,
                addend: 0,
                is_weak: false,
                start_of_sequence_offset: 0,
            },
            Import {
                name: "_printf",
                dylib: "/usr/lib/libSystem.B.dylib",
                is_lazy: true,
                offset: 4104,
                size: 8,
                address: 8200,
                addend: 0,
                is_weak: false,
                start_of_sequence_offset: 0,
            },
        ],
    ),
}
